### ISSUE TITLE
Providing Options to Transformations

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#478]
For my ROMS work I need to disable the test RegionTrans._validate to check for valid node types: if I want to apply an OMP parallel directive to statements that contains assignments (required since I sometimes have assignment between loop that should be in one parallel region), the _validate check prohibits this (since assignments are not allowed):
```
    valid_node_types = (psyGen.Loop, psyGen.Kern, psyGen.BuiltIn,
                        psyGen.OMPDirective, psyGen.GlobalSum,
                        psyGen.Literal, psyGen.Reference)
```
While I could easily add assignment to this list, I think it would be good to have a more flexible system of handling this.

Obviously we could provide a parameter to OMPParallelTrans constructor or apply() function (or a separate setter for valid node types) where we can modify the valid node types.

But I am wondering if it might be worth adding a more flexible 'transformation option' system for all transformations: each apply() function would take a dictionary, where additional options can be specified. So we could call for example:
```
omptrans.apply({"disable node type check":1}), 
or:
omptrans.apply({"additional_node_types":[psyGen.Assignment, psyGen.If] }), 
```
The advantage is that all apply() function would have the same signature (one dict, instead of: this transform takes the optional flags a, b, or c, and that transforms might take a and d ... you get the idea)  making it easy to call other transformations, since we just pass the option-dictionary from one apply() function to the next. While there is no real immediate benefit of doing it this way, it _might_ make it more easier long term to have much more complex transformations. A (made up) example:

assume we have a transform ''SuperTrans'' that first renames scalar variables (in case of different usage of the same variable, once thread private, then global) and moves them to the beginning, to allow loop fusion. Then we fuse loops, and apply openmp parallel and omp do as much as possible. Obviously ''SuperTrans'' would consist of several simpler transformations.

Then it might be the case that the omp transform can not decide if a loop can be parallelised (as a real example: assignments to a user defined type a%b(i,j)=... are currently not handled in nemo/PSyir, and I need to overwrite the handling of this special case for one loop). Since the user script only calls ''SuperTrans'', we can't immediately set any flags in the omp transformation.But if we would have one generic option dictionary, we could call ''SuperTrans'' with {'omp-parallelise-user-defined-type':1}. This would not have any affect to any other transform, but the omp transform can test for this setting, and if so enforce parallelisation of the loop.

Downside of course - if there should be a typo in a parameter name, it will not be detected (unless we add some complicated code to keep track of valid parameters ...).

It might be an overkill (or have other disadvantages that I am not thinking of atm). And there might be other solutions (provide ''SuperTrans'' with instances of the transformation it needs, so we can create customise the behaviour this way??, ...)

Any comments? And as a more direct question: if you don't like this approach, what is the best way for me to disable the validate check? I like the idea of the check, so I wouldn't suggest to remove it, or just add more in (since basically nearly everything might be allowed in a parallel region, so if we add all, the test becomes useless). Would you prefer a simple solution (just a flag for the omp transform to ignore the test, or a way to optionally add more allowed classes?

**Comment by hiker** [comment]
Discussion this issue with @rupertford, we came up with the following points:
- The _validate function should be called validate. This allows tools to do their own validation check without the transform (as a use case: auto-tuning or a potential GUI code need a list of valid transforms to apply).
- The OMP test should test which nodes are NOT supposed to be surrounded by OMP parallel, not list which nodes are allowed.
- A general way of adding options to transform is useful. Use case for a general approach (as opposed to transformation specific parameter for each transform): I see myself having complicated 'top level' transforms/scripts, which uses several other transforms (e.g. renaming scalar variables, then moving assignment to scalar variables up to allow loops to be fused, then loop fusion, then omp, ..). There could easily be a need that in the top-level user script I need to disable a test several levels down in the call tree: e.g. a loop might contain a call, which in general should not be parallelised, but in a special case I might know that the call is to a math function that can be done in parallel, so I want allow calls in this case. So I would want to pass in a flag to my top-level optimisation script, which will be passed on to the OMP parallel transform later. This is significantly simpler than adding various flags to the top level script, which trigger other flags in subsequent transforms. Also each script can add or remove options as required by adding/removing entries to the option object.

All of this should be done in different steps/pull requests (not necessarily in this order), so I'll open different tickets for renaming and the test in omp parallel.

**Comment by hiker** [comment]
See #492, #493.

**Comment by hiker** [comment]
#492 was a duplicate of #208 - thanks @TeranIvy

**Comment by hiker** [comment]
As summary of today's discussion: it would be possible to use 'set' and 'dictionary' notation for flags:
```
  apply(..., {"disable-node-type-check", "reprod"})            # set syntax
  apply(..., {"disable-node-type-check":True, "collapse":2})   # dictionary syntax
```
since most tests will just be: ```if "disable-node-type-check" in options: ``` (which works for set and dictionary syntax). But the moment one option needs a value (like collapse), every option must be written using dictionary syntax, (```{"a", "b":1}``` is invalid).

But we agreed to always use dictionary syntax, e.g.:
```
   apply(..., {"node-type-check": False})
```
For now we also won't add a kind of 'namespace' (e.g. "omp:collapse", "openacc:collapse" or so).

**Comment by hiker** [comment]
I found a few more issues we need to agree on:
1) I think we should define if a function accepts certain options (otherwise we actually lose documentation). My suggestion is to document them as parameter with the name 'options-name', e.g.:
```
:param bool options-const_bounds: whether the constant loop should be used (True) ...
```
Or maybe use '.' instead of '-'? Ideally ':', but I think this will interfere with spinx's parsing.

2) In some cases `apply` calls `validate` in the base class only (because there is no `validate` in the class), e.g.:
```
def apply(...)
   super(MyName, self).validate(...)
```
I am changing those to be `self.validate(...)` - is that ok? This way if we should ever add a `validate` method (or inherit from this class and provide `validate` there), the new `validate` will actually be called.

3) Not all `validate` methods call `validate` of the base class. Even though the base class might not do any validation, I think we should still put the call in (in case that we add more validation later)

4) If we agree to the above, we now have potentially too much validation happening. While it doesn't matter technically, it might cause some slowdown:
```
class A(Transformation)
   def validate(self, ...):
       super(A, self).validate(...)
   def apply(self, ...):
       self.validate(...)
        ...
class B(A):
   def validate(self):
        super(B, self).validate(...)
   def apply(self, ...):
        self.validate(...)
        super(B, self).apply(...)
```
If we call `B().apply()`, this will first call `B.validate()`, which will call `A.validate()`. Then it calls `A.apply`, which again will call `B.validate()` (and then `A.validate()` again). I can't find an obvious way around this ... except for adding an option 'has-been-validated',which will be tested before calling validation. So in each `apply()` we would need:
```
   def apply(self, ..., options)
        if not options.get("has-been-validated", False):
            self.validate(options)
            options = dict(options)   # Create a copy to avoid modifying the caller's options dict
            options["has-been-validated"] = True
```
This way even the caller can easily disable validation if ever required.

If we agree to the latter (i.e. no one has a better idea), then I would suggest to make this part of a independent pull request, to make it quicker to merge this in (which could cause quite a few conflicts).

@sergisiso , @rupertford, @TeranIvy, @TeranIvy  - feedback welcome.

**Comment by hiker** [comment]
One more issue: `dynamo0p3_transformation_test.test_reprod_reduction_real_pdo()` tests if an exception is raised if the invalid keyword 'reprod' is given to `DynamoOMPParallelLoopTrans.apply()`. The exception is raised by python, not by an explicit test anywhere.

This test fails now (since we don't have keywords anymore). Two options:
1) I explicitly test for 'reprod' in options in the corresponding validate() function. Though imho this kind of contradicts the spirit of the options dictionary - the idea that we can supply a list of options, and the transforms only checks for those options it supports.
2) We just remove the test.

**Comment by hiker** [comment]
One small (suggested) change:I found that using '-' in options look neater, e.g.:
`options["check-node-type"]`   (instead of using _). Since this then clashes with my suggested way of documenting the supported options, I propose:
```
        :param bool options["node-type-check"]: This flag controls if the \
                type of the nodes enclosed in the region should be tested.
```
This should be more intuitive to understand for users (and it also renders fine in html and pdf).

**Comment by hiker** [comment]
Reopening, since there is one outstanding issue that I missed: same_space in lfric_loop_fuse_trans.py:
